R/updateNullModelApproxScoreSE.R

```diff
@@ -0,0 +1,246 @@
+
+     # check for W matrix
+     if (is.null(null.model$W)) stop('This null model was created with an older version of GENESIS and is not compatible with this analysis; 
+          please re-run your null model with the latest version.')
+
+     # samples in null model
+     sampid <- null.model$fit$sample.id
+     if(verbose) message(paste('null.model has', length(sampid), 'samples'))
+
+     # select a random set of variants meeting min.mac in the sample set
+     varid <- .selectRandomVars(gdsobj, sample.id = sampid, nvar = nvar, min.mac = min.mac, verbose = verbose)
+     if(verbose) message(paste('Selected', length(varid), 'variants with MAC >=', min.mac))
+
+     # calculate Score.SE using both approaches
+     tab <- .scoreSEtable(gdsobj, null.model, variant.id = varid, 
+                         sparse = sparse, imputed = imputed, male.diploid = male.diploid, genome.build = genome.build, verbose = verbose)
+	
+     # calculate the score.se.ratio estimate
+     score.se.ratio <- .scoreSEratio(tab)
+
+     # update the null model
+     null.model <- nullModelSmall(null.model)
+     null.model$score.se.ratio <- score.se.ratio
+     return(null.model)
+}
+
+
+## function to get a random sample of variants with a minimum MAC
+## this is called by updateNullModelApproxScoreSE
+setGeneric(".selectRandomVars", function(gdsobj, ...) standardGeneric(".selectRandomVars"))
+
+setMethod(".selectRandomVars",
+          "SeqVarData",
+          function(gdsobj, sample.id = NULL, nvar = 100, min.mac = 20, verbose = TRUE){
+
+               # number of variants in the GDS object
+               nvar.gdsobj <- SeqArray::seqSummary(gdsobj, verbose = FALSE)$num.variant
+               if(nvar.gdsobj < nvar) stop('requested more variants than available in gdsobj')
+
+               # filter to sample set in null.model
+               SeqArray::seqSetFilter(gdsobj, sample.id = sample.id, verbose = verbose)
+
+               out <- NULL
+               while(length(out) < nvar){
+                    # sample variants
+                    var.rand <- sort(sample.int(nvar.gdsobj, size = nvar, replace = FALSE))
+                    # filter to random sample
+                    SeqArray::seqSetFilter(gdsobj, variant.sel = var.rand, verbose = FALSE)
+                    # filter to MAC threshold
+                    if(min.mac > 0) SeqArray::seqSetFilterCond(gdsobj, mac = min.mac, verbose = FALSE)
+                    # collect selected variants
+                    out <- sort(unique(c(out, seqGetData(gdsobj, 'variant.id'))))
+               }
+
+               # sample down to number requested
+               out <- sample(out, size = nvar, replace = FALSE)
+               return(out)
+               })
+
+setMethod(".selectRandomVars",
+          "GenotypeData",
+          function(gdsobj, sample.id = NULL, nvar = 100, min.mac = 20, verbose = TRUE){
+
+               stop("function not yet implemented for GenotypeData objects")
+
```

Comment:
I started trying to put this together for a GenotypeData object, but it seems more challenging, and I'm not sure if it's worthwhile.